tests/unit/test_loaders.py

```diff
@@ -126,3 +145,113 @@ def test_load_measure_definitions_empty_measures(funcs):
+        loaders.isolation_is_required({"EHRQL_ISOLATE_USER_CODE": "maybe"})
+
+
+@patch.object(loaders, "isolation_is_supported", return_value=False)
+def test_load_definition_raises_error_if_isolation_required_but_unavailable(_):
+    filename = FIXTURES_GOOD / "dataset_definition.py"
+    with pytest.raises(RuntimeError, match="current environment does not support"):
+        loaders.load_dataset_definition(
+            filename,
+            user_args=(),
+            environ={"EHRQL_ISOLATE_USER_CODE": "always"},
+        )
+
+
+def test_load_definition_unsafe_raises_error_if_isolation_required():
+    filename = FIXTURES_GOOD / "dataset_definition.py"
+    with pytest.raises(RuntimeError, match="call to unsafe loader function"):
+        loaders.load_definition_unsafe(
+            "dataset",
+            filename,
+            user_args=(),
+            environ={"EHRQL_ISOLATE_USER_CODE": "always"},
+        )
+
+
+# Confirm that various things we can do in an ordinary subprocess are blocked in an
+# isolated subprocess
+@pytest.mark.parametrize(
+    "function,status",
+    [
+        (subprocess.run, "OK"),
+        (loaders.subprocess_run_isolated, "BLOCKED"),
+    ],
+)
+@pytest.mark.skipif(
+    not sys.platform.startswith("linux"),
+    reason="Subprocess isolation only works on Linux",
+)
+def test_subprocess_run_isolated(tmp_path, function, status):
+    assert loaders.isolation_is_supported()
+    # Attempt various sorts of action and print whether or not we were blocked
+    code = textwrap.dedent(
+        """\
+        import os, pathlib, socket, subprocess
+
+
+        print("Touch: ", end="")
+        try:
+            pathlib.Path(".").touch()
+            print("OK")
+        except PermissionError:
+            print("BLOCKED")
+
+        print("Open socket: ", end="")
+        try:
+            socket.create_connection(("192.0.2.0", 53), timeout=0.001)
+        except TimeoutError:
+            print("OK")
+        except PermissionError:
+            print("BLOCKED")
+
+        print("Exec: ", end="")
+        try:
+            subprocess.run(["/bin/true"])
+            print("OK")
+        except PermissionError:
+            print("BLOCKED")
+
+        print("Read env vars: ", end="")
+        try:
+            pathlib.Path(f"/proc/{os.getppid()}/environ").read_bytes()
+            print("OK")
+        except PermissionError:
+            print("BLOCKED")
+        """
+    )
+    result = function(
+        [sys.executable, "-c", code],
+        check=True,
+        text=True,
+        stdout=subprocess.PIPE,
+        cwd=tmp_path,
+    )
+    assert result.stdout == (
+        f"Touch: {status}\n"
+        f"Open socket: {status}\n"
+        f"Exec: {status}\n"
+        # Until we can assume "unveil" support we can't block the subprocess reading its
+        # parent's environment variables out of /proc; but we leave this test code in
+        # place in preparation.
+        f"Read env vars: OK\n"
+    )
```

Comment:
Could we turn this test into a subcommand that we can also run inside the production system, to verify that the protections are enabled there?

Reply:
Yes, good suggestion. Thanks.